### PR TITLE
[Relay] Add grads

### PR DESCRIPTION
**Pull request by MarisaKirisame**
This is the continuation of #3443 since @altanh is gone for summer intern.

### CODE REVIEW

**Comment by jroesch** [comment]
Can you just add some tests and we can merge?

**Comment by MarisaKirisame** [comment]
@jroesch yes. I will remove the take, and I think for now, the easiest way will be to get a take_grad.
the gradient of take_grad will have the great property of being all zero.

File: python/tvm/relay/op/_tensor_grad.py [hunk 1]
```diff
@@ -130,3 +137,137 @@ def clip_grad(orig, grad):
...
+    # batchnorm has a wrong api so we will not waste time implementing it.
+    a, b, c, d, e = orig.args
+    return [a, b, c, d, e]
+
+
+@register_gradient("nn.conv2d")
+def conv2d_grad(orig, grad):
+    a, b = orig.args
+    x, y = relay.TupleWrapper(relay.Call(op.get("nn.conv2d_grad"), [a, b, grad], orig.attrs), 2)
+    return [a, b]
```
Comment (SWu) [inline python/tvm/relay/op/_tensor_grad.py]: Which of the grad implementations here are completed? A lot of these look like placeholders; e.g. the conv2d and pool2d grads aren't returning real gradients here.

Comment (MarisaKirisame) [reply python/tvm/relay/op/_tensor_grad.py]: basically only entropy. this is mostly a WIP.

**Comment by tqchen** [comment]
@MarisaKirisame please act on the CI problem and comments

**Comment by MarisaKirisame** [comment]
@tqchen there's alot of todo rn, and I am on them. I think I should split this into small chunk.

File: python/tvm/relay/op/_tensor_grad.py (hunk 1 above)
Comment (sgrechanik-h) [reply python/tvm/relay/op/_tensor_grad.py]: It would be better to mark incomplete gradients as incomplete somehow, e.g. using a lower `plevel`.

Comment (MarisaKirisame) [reply python/tvm/relay/op/_tensor_grad.py]: @sgrechanik-h it is ok, this is mostly for testing rn. I think I will open lots of small PR for the gradients.

File: python/tvm/relay/op/nn/_nn.py [hunk 2]
```diff
@@ -548,3 +548,13 @@ def schedule_deformable_conv2d(attrs, outs, target):
...
+
+reg.register_schedule("nn.cross_entropy", schedule_injective)
+
+reg.register_pattern("nn.cross_entropy",
+                     OpPattern.OPAQUE)
+
+@reg.register_compute("nn.cross_entropy")
+def compute_cross_entropy(attrs, inputs, out_dtype, target):
+    x, y = inputs
+    return [-topi.sum(topi.nn.log_softmax(x) * y / x.shape[0])]
```
Comment (SWu) [inline python/tvm/relay/op/nn/_nn.py]: This looks like it's doing fused softmax+cross-entropy from logits. For a normal feed-forward model that's converted into relay, the outputs will already be softmax; do you have an example of how you'd remove that softmax layer and add this cross-entropy from logits layer instead?

Comment (MarisaKirisame) [reply python/tvm/relay/op/nn/_nn.py]: I havent thought that much when I was doing this - I am only trying to get an end to end training example.

**Comment by MarisaKirisame** [comment]
Closing because too big and outdated. will reopen as multiple small pr.
